src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -3153,6 +3154,105 @@ protected override void HandleSetInstructionBreakpointsRequestAsync(IRequestResp
+            }
+            else
+            {
+                success = m_frameHandles.TryGet(frameId, out frame);
+            }
+
+            if (!success)
+            {
+                Dictionary<string, object> properties = new Dictionary<string, object>();
+                properties.Add(DebuggerTelemetry.TelemetryStackFrameId, frameId);
+                DebuggerTelemetry.ReportError(DebuggerTelemetry.TelemetryEvaluateEventName, 1108, "Invalid frameId", properties);
+                responder.SetError(new ProtocolException("Cannot evaluate expression on the specified stack frame."));
+                return;
+            }
+
+            uint radix = Constants.EvaluationRadix;
+
+            if (responder.Arguments.Format != null)
+            {
+                ValueFormat format = responder.Arguments.Format;
+
+                if (format.Hex == true)
+                {
+                    radix = 16;
+                }
+            }
+
+            if (m_settingsCallback != null)
+            {
+                // MIEngine generally gets the radix from IDebugSettingsCallback110 rather than using the radix passed
+                m_settingsCallback.Radix = radix;
+            }
+
+            var eb = new ErrorBuilder(() => string.Format(CultureInfo.CurrentCulture, AD7Resources.Error_SetExpression, expression, value));
+
+            try
+            {
+
+                IDebugExpressionContext2 expressionContext;
+                int hr = frame.GetExpressionContext(out expressionContext);
+
+                IDebugExpression2 expressionObject;
+                string error;
+                uint errorIndex;
+                hr = expressionContext.ParseText(expression, enum_PARSEFLAGS.PARSE_EXPRESSION, Constants.ParseRadix, out expressionObject, out error, out errorIndex);
+                eb.CheckHR(hr);
+                if (!string.IsNullOrEmpty(error))
+                {
+                    DebuggerTelemetry.ReportError(DebuggerTelemetry.TelemetryEvaluateEventName, 4001, "Error parsing expression");
+                    responder.SetError(new ProtocolException(error));
+                    return;
+                }
+
+                // NOTE: This is the same as what vssdebug normally passes for the watch window
+                enum_EVALFLAGS flags = enum_EVALFLAGS.EVAL_RETURNVALUE |
+                    enum_EVALFLAGS.EVAL_NOEVENTS |
+                    (enum_EVALFLAGS)enum_EVALFLAGS110.EVAL110_FORCE_REAL_FUNCEVAL;
+
+                IDebugProperty2 property;
+                hr = expressionObject.EvaluateSync(flags, Constants.EvaluationTimeout, null, out property);
+                eb.CheckHR(hr);
+                eb.CheckOutput(property);
+
+                hr = property.SetValueAsString(value, radix, Constants.EvaluationTimeout);
+                eb.CheckHR(hr);
+
+
+                responder.SetResponse(new SetExpressionResponse
+                {
+                    Value = value
```

Comment:
>value [](http://example.com/codeflow?start=28&length=5)

We should probably re-query the value since the format of the user-provided text doesn't necessarily match what GDB will return. Trivial example: `myIntVar = 0x12` when hex mode is disabled.